util/src/matrix.rs

```diff
@@ -0,0 +1,527 @@
+    ) -> impl Iterator<Item = &mut [T]> + DoubleEndedIterator + ExactSizeIterator + FusedIterator
+    {
+        let width = self.width;
+        self.as_flat_mut().chunks_mut(width)
+    }
+
+    /// Transpose the matrix.
+    ///
+    /// This function may re-use memory but is not required to do so.
+    pub fn transpose(self) -> Self {
+        let height = self.height;
+        let width = self.width;
+
+        let initializer_fn = |res: &mut Matrix<MaybeUninit<T>>| {
+            // We're essentially moving all contents of `self` to `res`. We need to ensure that they
+            // don't get dropped when `self` gets dropped, so we wrap them in `ManuallyDrop`.
+            let mut me = self.manually_drop_elements();
+
+            // Important that we don't panic here. We've disabled the destructors of all our
+            // elements, so they won't be called if we have to unwind, potentially causing a leak.
+
+            // Easy optimization: ensure the larger loop is outside.
+            unsafe {
+                if width >= height {
+                    for i in 0..width {
+                        for j in 0..height {
+                            // SAFETY: The indices are valid. The pointee of `in_ref` is not used
+                            // again (so we're effectively just moving it).
+                            // idk why `in_ref` has to be `mut`, but `ManuallyDrop::take` requires
+                            // it.
+                            let in_ref = me.get_unchecked_mut(j).get_unchecked_mut(i);
+                            let out_ref = res.get_unchecked_mut(i).get_unchecked_mut(j);
+                            out_ref.write(ManuallyDrop::take(in_ref));
+                        }
+                    }
+                } else {
+                    for j in 0..height {
+                        for i in 0..width {
+                            // SAFETY: see above.
+                            let in_ref = me.get_unchecked_mut(j).get_unchecked_mut(i);
+                            let out_ref = res.get_unchecked_mut(i).get_unchecked_mut(j);
+                            out_ref.write(ManuallyDrop::take(in_ref));
+                        }
+                    }
+                };
+            }
+        };
+
+        unsafe {
+            // SAFETY: The closure initializes every element.
+            Self::new_with(width, height, initializer_fn)
+        }
+    }
+
+    /// Transmute each element of the matrix from `T` to `U`.
+    ///
+    /// # Safety
+    ///
+    /// Due to interactions with the allocator, this method has stricter requirements than the
+    /// `mem::transmute` function in the library.
+    ///
+    /// Same as `mem::transmute`:
+    /// - `T` and `U` must have the same size.
+    /// - All elements of the matrix must be safe to interpret bitwise as `U`.
+    ///
+    /// Furthermore:
+    /// - `T` and `U` must have the same alignment.
+    unsafe fn transmute<U>(self) -> Matrix<U> {
```

Comment:
I'm probably missing something but can this use `mem::transmute` (to implement this)?

Reply:
Not directly, because Rust doesn't actually guarantee how `Matrix` is laid out in memory. Although we could force a particular layout with `#[repr(C)]`.